### ISSUE TITLE
Add arithmetic operators to valarray

### ISSUE THREAD

**Issue by kylelutz** [boostorg/compute#381]
Add overloads for the arithmetic operators (e.g. `+`, `-`) to `boost::compute::valarray<T>` to match `std::valarray<T>`.

From the review:
```
- <http://kylelutz.github.io/compute/boost/compute/valarray.html> -
presumably compute::valarray has overloaded arithmetic operators?  Are
they mentioned anywhere?
```

**Comment by guptadivyansh** [comment]
http://www.cplusplus.com/reference/valarray/valarray/operators/

Similar to these?

**Comment by kylelutz** [comment]
@guptadivyansh yeah, those are the ones.

**Comment by guptadivyansh** [comment]
https://github.com/guptadivyansh/compute/commit/1771ad07b9dd3f893e2497f6b17ece46ee391c85

@kylelutz Can you please look at this?

**Comment by kylelutz** [comment]
So that will actually perform the operation on the host rather than on the compute device (e.g. GPU). For implementing these functions I'd probably use the [`boost::compute::transform()`](http://kylelutz.github.io/compute/boost/compute/transform.html) algorithm.

**Comment by archangel4** [comment]
Could you provide some more reference to implement this feature ?

**Comment by kylelutz** [comment]
A basic implementation would be something like:
```
valarray<T> operator+(const valarray<T> &other) const
{
    const context &context = m_buffer.get_context();
    command_queue queue(context, context.get_device());
    valarray<T> result(size(), context);
    transform(begin(), end(), other.begin(), plus<T>(), queue);
    queue.finish();
    return result;
}
```

As operator overloads don't allow the command queue argument to be passed, we must create one explicitly to perform the operation. Because of that, actually using these operators would not be ideal for performance. For that reason I'd be hesitant to add them, but they could be useful for their convenience.

**Comment by haahh** [comment]
Logical not operator ``valarray<bool> operator!() const;`` cannot be implemented as in OpenCL it's impossible to build memory buffer with bool type, which makes valarray<bool> incorrect. However, I can implement ``valarray<char> operator!() const;`` operator (or with int) .

**Comment by kylelutz** [comment]
Yeah, I'd just use `valarray<char>`.

**Comment by kylelutz** [comment]
Implemented in PR #454.
